src/libsyntax/parse/parser.rs

```diff
@@ -2030,15 +2034,20 @@ impl<'a> Parser<'a> {
     }
 
     /// Parse single lifetime 'a or panic.
-    fn expect_lifetime(&mut self) -> Lifetime {
-        match self.token {
-            token::Lifetime(ident) => {
-                let ident_span = self.span;
-                self.bump();
-                Lifetime { ident: ident, span: ident_span, id: ast::DUMMY_NODE_ID }
+    pub fn expect_lifetime(&mut self) -> Lifetime {
+        let lifetime = match self.token {
+            token::Lifetime(ident) =>
+                Lifetime { ident: ident, span: self.span, id: ast::DUMMY_NODE_ID },
+            token::Interpolated(ref nt) => match nt.0 {
+                token::NtLifetime(lifetime) =>
+                    lifetime,
+                _ => self.span_bug(self.span, "not a lifetime")
```

Comment:
Optional improvement -- to avoid duplicating these matches, I'd refactor out `token.lifetime()` such that `token.is_lifetime()` is `token.lifetime().is_some()`; see `token.ident()`.

Reply:
That sounds nice but it isn't that straightforward, since token::Lifetime has an ident and token::NtLifetime has a lifetime.

Reply:
I see. I think it would work if `token.lifetime()` took an argument `token.lifetime(span)` so that the `token::Lifetime` case would return `Lifetime { ident, span, id: ast::DUMMY_NODE_ID }`.